### PR TITLE
[ML] Add mixed cluster tests for inference

### PR DESCRIPTION
**Pull request by maxhniebergall**
Added mixed cluster tests for the inference service. I copied the integration tests from the Upgrade tests, and I copied the MixedCluster set up [from this PR](https://github.com/elastic/elasticsearch/pull/102198)

## To run the new mixed cluster tests
this command can be used: `./gradlew ':x-pack:plugin:inference:qa:mixed-cluster:v8.15.0#javaRestTest'`. However, to run the mixed cluster tests against main (rather than a previously released version), the code base must be changed, and this flag must be added to the gradlew command ` -Dtests.bwc.refspec.8.15=origin/main`. The code changes required to run bwc-mixed-cluster tests against main have been added to TESTING.asciidoc and are:

`Another use case, since the introduction of serverless, is to test BWC against main in addition to the other released branches. To do so, specify the bwc.refspec remote and branch to use for the BWC build as origin/main. To test against main, you will also need to create a new version in [Version.java](https://github.com/elastic/elasticsearch/pull/server/src/main/java/org/elasticsearch/Version.java), increment elasticsearch in [version.properties](https://github.com/elastic/elasticsearch/pull/build-tools-internal/version.properties), and hard-code the project.version for ml-cpp in [ml/build.gradle](https://github.com/elastic/elasticsearch/pull/x-pack/plugin/ml/build.gradle).`

## Manual testing
I tested making changes to the serialization `TextEmbeddingResults` and only syntax errors were detected as failures, adding new fields didn't cause failures (I'll have to think more about whether this makes sense or not).

I removed `CohereEmbeddingsTaskSettings` from the `InferenceNamedWriteablesProvider` and this caused the `CohereServiceMixedIT` to fail, which is great!

I added a fake field to `CohereEmbeddingServiceSettings StreamInput` and the test failed. I added a new field to `CohereEmbeddingServiceSettings toXContent` and the test succeeded.  I added a new field to `CohereEmbeddingServiceSettings writeTo` and the test succeeded.

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/RestrictedBuildApiService.java [hunk 1]
```diff
@@ -128,6 +128,7 @@ private static ListMultimap<Class<?>, String> createLegacyRestTestBasePluginUsag
         map.put(LegacyRestTestBasePlugin.class, ":x-pack:plugin:ilm:qa:multi-cluster");
         map.put(LegacyRestTestBasePlugin.class, ":x-pack:plugin:ilm:qa:multi-node");
         map.put(LegacyRestTestBasePlugin.class, ":x-pack:plugin:ilm:qa:rest");
+        map.put(LegacyRestTestBasePlugin.class, ":x-pack:plugin:inference:qa:mixed-cluster");
```
Comment (dnhatn) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/RestrictedBuildApiService.java:131]: I think we don't need this since we are using the new test framework?

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/AzureOpenAiServiceMixedIT.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        openAiEmbeddingsServer.start();
+    }
+
+    // @AfterClass test mute
+    public static void shutdown() {
+        openAiEmbeddingsServer.close();
+    }
+
+    @SuppressWarnings("unchecked")
+    @LuceneTestCase.AwaitsFix(bugUrl = "AzureOpenAI doesn't support webserver URL in tests")
```
Comment (mark-vieira) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/AzureOpenAiServiceMixedIT.java:46]: This should point to an open GitHub issue. If we never intend to run this test then why bother having it?

File: x-pack/plugin/inference/qa/mixed-cluster/build.gradle [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import org.elasticsearch.gradle.VersionProperties
+import org.elasticsearch.gradle.util.GradleUtils
+import org.elasticsearch.gradle.internal.info.BuildParams
+import org.elasticsearch.gradle.testclusters.StandaloneRestIntegTestTask
+
+apply plugin: 'elasticsearch.internal-java-rest-test'
+apply plugin: 'elasticsearch.internal-test-artifact-base'
+apply plugin: 'elasticsearch.bwc-test'
+
+restResources {
```
Comment (mark-vieira) [inline x-pack/plugin/inference/qa/mixed-cluster/build.gradle:11]: This block is only used for yaml tests but I see none here.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/RestrictedBuildApiService.java (hunk 1 above)
Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/RestrictedBuildApiService.java:131]: Correct. This should be unnecessary.

File: x-pack/plugin/inference/build.gradle [hunk 4]
```diff
@@ -27,6 +27,7 @@ base {
 
 dependencies {
   implementation project(path: ':libs:elasticsearch-logging')
+  testImplementation project(path: ':test:framework')
```
Comment (mark-vieira) [inline x-pack/plugin/inference/build.gradle:30]: Why was this added? This should be automatically added to all projects. Plus this PR is just adding a new QA project.

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+
+public class BaseMixedIT extends ESRestTestCase {
```
Comment (mark-vieira) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java:33]: Should this be `abstract`? Otherwise if common tests get added here we'll attempt to run this as it's own test suite.

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/Clusters.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+package org.elasticsearch.xpack.inference.qa.mixed;
+
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.local.distribution.DistributionType;
+import org.elasticsearch.test.cluster.util.Version;
+
+public class Clusters {
+    public static ElasticsearchCluster mixedVersionCluster() {
+        Version oldVersion = Version.fromString(System.getProperty("tests.old_cluster_version"));
+        return ElasticsearchCluster.local()
```
Comment (mark-vieira) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/Clusters.java:17]: Is it strictly required that these mixed clusters be 4 nodes? It just makes them very heavyweight and expensive to run.

File: x-pack/plugin/inference/build.gradle (hunk 4 above)
Comment (maxhniebergall) [reply x-pack/plugin/inference/build.gradle:30]: I'll try removing it

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/Clusters.java (hunk 6 above)
Comment (maxhniebergall) [reply x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/Clusters.java:17]: I think 2 nodes should be enough, I'll decrease it

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        return format("http://%s:%s", webServer.getHostName(), webServer.getPort());
+    }
+
+    @Override
+    protected Settings restClientSettings() {
+        String token = basicAuthHeaderValue("x_pack_rest_user", new SecureString("x-pack-test-password".toCharArray()));
+        return Settings.builder().put(ThreadContext.PREFIX + ".Authorization", token).build();
+    }
+
+    protected void delete(String inferenceId, TaskType taskType) throws IOException {
```
[the remark this answers is not in the document]
Comment (maxhniebergall) [reply x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java:58]: Hmm, I'm not really sure how that would work. Where would this new class go?

[the remark this answers is not in the document]
Comment (maxhniebergall) [reply x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java:58]: I tried to do this, but it wasn't easy or obvious to me how to set up a new module to contain this class and to be able to access the dependencies

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/AzureOpenAiServiceMixedIT.java (hunk 2 above)
Comment (maxhniebergall) [reply x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/AzureOpenAiServiceMixedIT.java]: I removed this test for now, we will probably want to add it back once we get it working with the webserver.

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/Clusters.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            .distribution(DistributionType.DEFAULT)
+            .withNode(node -> node.version(oldVersion))
+            .withNode(node -> node.version(Version.CURRENT))
+            .setting("xpack.security.enabled", "false")
+            .setting("xpack.license.self_generated.type", "trial")
+            .setting("cluster.routing.rebalance.enable", "none") // disable relocation until we have retry in ESQL
```
Comment (davidkyle) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/Clusters.java:23]:
```suggestion
```

File: x-pack/plugin/inference/qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/xpack/application/OpenAiServiceUpgradeIT.java [hunk 9]
```diff
@@ -120,6 +120,7 @@ public void testOpenAiCompletions() throws IOException {
         final String upgradedClusterId = "upgraded-cluster-completions";
 
         if (isOldCluster()) {
+            // TODO why is put only in old cluster?
```
Comment (davidkyle) [inline x-pack/plugin/inference/qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/xpack/application/OpenAiServiceUpgradeIT.java:123]:
```suggestion
```

There is another endpoint put in the upgraded cluster at line 148. However, there is no put in the mixed cluster, is that what you are referring to?

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/CohereServiceMixedIT.java [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        var inferenceMap = inference(inferenceId, TaskType.TEXT_EMBEDDING, "some text");
+        assertThat(inferenceMap.entrySet(), not(empty()));
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testRerank() throws IOException {
+        var rerankSupported = bwcVersion.onOrAfter(Version.fromString(COHERE_RERANK_ADDED));
+        assumeTrue("Cohere rerank service added in " + COHERE_RERANK_ADDED, rerankSupported);
+
+        final String oldClusterId = "old-cluster-rerank";
```
Comment (davidkyle) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/CohereServiceMixedIT.java:135]: nit: there is no old cluster only a mixed cluster

```suggestion
        final String inferenceId = "mixed-cluster-cohere-rerank";
```

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/CohereServiceMixedIT.java [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        assertThat(embeddingType, Matchers.is(oneOf("int8", "byte")));
+
+        configs = (List<Map<String, Object>>) get(TaskType.TEXT_EMBEDDING, oldClusterIdFloat).get("endpoints");
+        serviceSettings = (Map<String, Object>) configs.get(0).get("service_settings");
+        assertThat(serviceSettings, hasEntry("embedding_type", "float"));
+
+        assertEmbeddingInference(oldClusterIdInt8, CohereEmbeddingType.BYTE);
+        assertEmbeddingInference(oldClusterIdFloat, CohereEmbeddingType.FLOAT);
+
+        {
```
Comment (davidkyle) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/CohereServiceMixedIT.java:85]: These 2 sections appear to be a repeat of the above with a different id. I can see the difference, if it is necessary please add a comment explaining what I am missing. 

Again the names `oldClusterId...` and `upgradedClusterId...` don't apply in the mixed cluster test as there is only the mixed cluster

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+import org.elasticsearch.test.http.MockWebServer;
+import org.elasticsearch.test.rest.ESRestTestCase;
+import org.hamcrest.Matchers;
+import org.junit.ClassRule;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public abstract class BaseMixedIT extends ESRestTestCase {
```
Comment (davidkyle) [inline x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/BaseMixedIT.java:30]: Typically when a class doesn't contain any tests the name ends in `TestCase` not `IT`. So `BaseMixedTestCase` for this class

**Review by davidkyle — COMMENTED** [review]
Looks good, thanks for adding these tests I left some minor comments.

I am confused that the concrete test classes inherit from `BaseMixedIT` not `MixedClusterSpecIT` as the `MixedClusterSpecIT` has the mixed cluster from Clusters.java. Should `BaseMixedIT` be using `public static ElasticsearchCluster cluster = Clusters.mixedVersionCluster();`

File: x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/CohereServiceMixedIT.java (hunk 11 above)
Comment (maxhniebergall) [reply x-pack/plugin/inference/qa/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/inference/qa/mixed/CohereServiceMixedIT.java:85]: Thanks for catching this!

**Comment by maxhniebergall** [comment]
> I am confused that the concrete test classes inherit from `BaseMixedIT` not `MixedClusterSpecIT` as the `MixedClusterSpecIT` has the mixed cluster from Clusters.java. Should `BaseMixedIT` be using `public static ElasticsearchCluster cluster = Clusters.mixedVersionCluster();`

Thanks for calling this out! It seems that this hierarchy mixup was hiding some underlying problems.

File: TESTING.asciidoc [hunk 13]
```diff
@@ -551,13 +551,19 @@ When running `./gradlew check`, minimal bwc checks are also run against compatib
...
-Sometimes a backward compatibility change spans two versions. A common case is a new functionality
-that needs a BWC bridge in an unreleased versioned of a release branch (for example, 5.x).
-To test the changes, you can instruct Gradle to build the BWC version from another remote/branch combination instead of
-pulling the release branch from GitHub. You do so using the `bwc.remote` and `bwc.refspec.BRANCH` system properties:
+Sometimes a backward compatibility change spans two versions.
+A common case is a new functionality that needs a BWC bridge in an unreleased versioned of a release branch (for example, 5.x).
+Another use case, since the introduction of serverless, is to test BWC against main in addition to the other released branches.
+To do so, specify the `bwc.refspec` remote and branch to use for the BWC build as `origin/main`.
+To test against main, you will also need to create a new version in link:./server/src/main/java/org/elasticsearch/Version.java[Version.java], 
+increment `elasticsearch` in link:./build-tools-internal/version.properties[version.properties], and hard-code the `project.version` for ml-cpp
```
Comment (jonathan-buttner) [inline TESTING.asciidoc:559]: Just curious why we need to change the version in ml cpp?

**Review by jonathan-buttner — APPROVED** [review]

File: TESTING.asciidoc (hunk 13 above)
Comment (maxhniebergall) [reply TESTING.asciidoc:559]: If you increment the elasticsearch version to one that doesn't exist, then the build system will not find the (non-existent) ML-CPP artifact. If, instead, we hardcode the current version of elasticsearch, the existing ML-CPP artifact will be downloaded.
